File: answers/leetcode/Coin Change/Coin Change.cpp
class Solution {
public:
    int coinChange(vector<int>& coins, int amount) {
        vector<int> dp(amount + 1, INT_MAX);
        dp[0] = 0;
        for (size_t i = 1; i < dp.size(); ++ i) {
            for (size_t j = 0; j < coins.size(); ++ j) {
                if (i >= coins[j] && dp[i - coins[j]] < INT_MAX) {
                    dp[i] = min(dp[i], dp[i - coins[j]] + 1);
                }
            }
        }
        return dp[amount] < INT_MAX ? dp[amount] : -1;
    }
};

File: answers/leetcode/Coin Change/Coin Change.java
class Solution {
    public int coinChange(int[] coins, int amount) {
        long[] dp = new long[amount + 1];
        Arrays.fill(dp, Long.MAX_VALUE);
        dp[0] = 0;
        for (int i = 0; i <= amount; i++) {
            if (dp[i] < Long.MAX_VALUE) {
                for (int j = 0; j < coins.length; j++) {
                    if ((long) i + (long) coins[j] <= (long) amount) {
                        dp[i + coins[j]] = Math.min(dp[i + coins[j]], dp[i] + 1l);
                    }
                }
            }
        }
        return (int) dp[amount];
    }
}